TypeCobol.LanguageServer/TypeCobolCustomLanguageServer/TypeCobolCustomLanguageServer.cs

```diff
@@ -23,29 +24,35 @@ public TypeCobolCustomLanguageServer(IRPCServer rpcServer, Queue<MessageActionWr
         /// </summary>
         public bool UseOutlineRefresh { get; set; }
 
-        protected override void OnShutdown()
-        {
-            if (UseOutlineRefresh && this.Workspace.DocumentModifiedEvent != null)
-                this.Workspace.DocumentModifiedEvent -= DocumentModified;
+        /// <summary>
+        /// Are we using the CFG view in the client.    
+        /// </summary>
+        public bool UseCfgDfaDataRefresh { get; set; }
 
-            base.OnShutdown();
+        protected override InitializeResult OnInitialize(InitializeParams parameters)
+        {
+            var result = base.OnInitialize(parameters);
+            this.Workspace.DocumentModifiedEvent += DocumentModified;
+            this.Workspace.ClientConfigurationChangedEvent += OnClientOptionsChanged;
+            return result;
         }
 
-        protected override void OnDidOpenTextDocument(DidOpenTextDocumentParams parameters)
+        /// <summary>
+        /// Handle -ol and -cfg options from the client configuration change notification.
+        /// </summary>
+        /// <param name="sender">Should be the Workspace instance</param>
+        /// <param name="options">Client's Options</param>
+        private void OnClientOptionsChanged(object sender, IEnumerable<string> options)
         {
-            if (UseOutlineRefresh && this.Workspace.DocumentModifiedEvent == null)
-            {
-                this.Workspace.DocumentModifiedEvent += DocumentModified;
-            }
-            base.OnDidOpenTextDocument(parameters);
+            //this.UseOutlineRefresh = options.Contains("-ol");
+            this.UseCfgDfaDataRefresh = options.Contains("-cfg");
         }
 
-        protected override void OnDidCloseTextDocument(DidCloseTextDocumentParams parameters)
+        protected override void OnShutdown()
         {
-            base.OnDidCloseTextDocument(parameters);
-            if (UseOutlineRefresh && this.Workspace.DocumentModifiedEvent != null && this.Workspace.IsEmpty)
-                this.Workspace.DocumentModifiedEvent -= DocumentModified;
-
+            this.Workspace.DocumentModifiedEvent -= DocumentModified;
+            this.Workspace.ClientConfigurationChangedEvent -= OnClientOptionsChanged;
+            base.OnShutdown();
         }
 
         protected override void OnDidSaveTextDocument(DidSaveTextDocumentParams parameters)
```

Comment:
hum.. it is safe to do that if the server is shuting down then OnDidReceiveRefreshOutline and UpdateCfgDfaInformation willnot be called, this the behavior in this PR.

Reply:
I don't get it, if the server is shuting down, there is no need to send Outline of CFG/DFA refresh.

Reply:
Yes, because there can still be pending parsing in process, so refresh event can come afterwards. Do you understand?

Reply:
Ok so the scenario is that we already unsubscribed from DocumentModified event (because of shutdown) and we have a pending refresh. I agree that the outline and cfg data will not be sent to client but
- why should the data be transmitted if we are terminating the server ?
- how is it related to OnDidSaveTextDocument ?

My idea is that `TypeCobolCustomLanguageServer` does not need to send any data about outline/cfg on a save operation because it is handled asynchronously with the refresh and consequently this override should be deleted.

Reply:
I don't remember the whole mechanism, but the save action is a way to resynchronize all data properly because sometimes incremental mode is out of sync.

Anyway, as the method `OnDidSaveTextDocument` is not in the scope of this PR, I suggest to discuss about it in another issue.

Reply:
I thought this would be included in this PR because @mayanje reviewed the outline mechanism along with the new cfg/dfa notification.

Reply:
The main thing is that DocumentModified event comes from another thread.

Reply:
New issue #1882 created.